ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/JmsXAResourceRecoveryHelper.java

```diff
@@ -82,31 +71,31 @@ public boolean initialise(String properties) {
      */
     @Override
     public XAResource[] getXAResources() {
-        if (connect()) {
-            if (jtaLogger.logger.isTraceEnabled()) {
-                jtaLogger.logger.trace("Returning XA resource: " + this);
+        if (!connectionManager.isConnected()) {
+            try {
+                connectionManager.connect();
+            } catch (XAException ignored) {
+                return new XAResource[0];
             }
-
-            return new XAResource[] { this };
         }
 
-        return new XAResource[0];
+        return new XAResource[] { this };
     }
 
     /**
      * Delegates XAResource#recover call to the connected JMS resource. If provided argument is XAResource.TMENDRSCAN, then JMS
      * connection will be closed at the end of the call.
      *
-     * @param i
+     * @param flag
      * @throws XAException
      */
     @Override
-    public Xid[] recover(int i) throws XAException {
+    public Xid[] recover(int flag) throws XAException {
         try {
-            return delegate.recover(i);
+            return connectionManager.connectAndApply(delegate -> delegate.recover(flag));
         } finally {
-            if (i == XAResource.TMENDRSCAN) {
-                disconnect();
+            if (flag == XAResource.TMENDRSCAN) {
+                connectionManager.disconnect();
```

Comment:
Based on your answer I am ok with the change so marking as approved. Is it possible to capture the original stacktrace in the JBTM report?